src/rascal/math/kvec_generator.hh

```diff
@@ -0,0 +1,137 @@
+       * - the Eigen vector of size N containing the norms of the k-vectors
+       */
+      Matrix_t kvectors{};
+      Vector_t kvector_norms{};
+
+     public:
+      /** Constructor:
+       * Upon initialization, directly compute all the k-vectors on the
+       * lattice specified by three basisvectors lying within a ball.
+       * @param cutoff Cutoff radius of reciprocal space ball.
+       *     As a guideline, if the smearing parameter of the density
+       *     is given by \f$sigma\f$, use \f$pi/sigma\f$ as cutoff.
+       * @param basisvectors 3x3 matrix containing the three basis
+       *     vectors a1, a2, a3 of the real space unit cell, stored as:
+       *     a1 = basisvectors[0], a2=basisvectors[1], etc.
+       *     These are then automatically converted to the corresponding
+       *     reciprocal cell vectors b1, b2, b3 by the formulae:
+       *     \f{equation}{b_1 = 2\pi \frac{a_1 \times a_2}
+       *                 {a_1 \cdot (a_2 \times a_3)} \f}
+       *     and similarly for the periodic permutations.
+       *     It is also possible to provide the basis vectors of the
+       *     reciprocal lattice directly by using the option below.
+       * @param is_reciprocal_cell false by default.
+       *     If set to true, it indicates that the cell provided
+       *     in the variable basisvectors is already the reciprocal cell.
+       *     Thus, the formula above for converting the a_i's to the b_i's
+       *     is not used, and the provided cell is used directly.
+       * @param need_origin false by default.
+       *     If set to true, the list of found k-vectors will also
+       *     contain the origin k=(0,0,0), which will then be stored
+       *     as the very first element in the matrix of found vectors.
+       */
+      Kvectors(double cutoff, Eigen::Matrix3d basisvectors,
+               bool is_reciprocal_cell = false, bool need_origin = false);
+
+      /** Returns number of vectors found within cutoff radius
+       * without double counting pairs related by inversion.
+       * This is the actual number that can be used to iterate over all
+       * vectors and its norms obtained from get_kvectors and get_norms.
+       */
+      size_t get_numvectors() const { return this->kvector_norms.size(); }
+
+      /** Returns const reference to the Eigen matrix containing all
+       * k-vectors within cutoff radius without double counting, where
+       * row(i) = i-th vector. If the origin k=(0,0,0) is included, it is
+       * always stored in row(0).
+       */
+      Matrix_t get_kvectors() const {
+        return Eigen::Ref<const Matrix_t>(this->kvectors);
+      }
+
+      /** Returns const reference to the Eigen vector containing the norm
+       * of all k-vectors within the cutoff without double counting.
+       */
+      Vector_t get_kvector_norms() const {
+        return Eigen::Ref<const Vector_t>(this->kvector_norms);
+      }
```

Comment:
You need to also change the return type of these functions. Here you are creating a const reference, which is then converted to a `Matrix_t`/`Vector_t` value, making a copy.